### PR TITLE
feat: Added the new search endpoint

### PR DESCRIPTION
**Pull request by dtoulcher-ch**
`/internal/users/search/` has been added for internal use
  allowing for partial serches on email addresses. The limit on the
  number of users returned is 50, but this can be configured.

CC-1248

__Short description outlining key changes/additions__

__JIRA Ticket Number__

### Type of change

* [ ] Bug fix
* [ ] New feature
* [ ] Breaking change
* [ ] Infrastructure change

### Pull request checklist

* [ ] I have added unit tests for new code that I have added
* [ ] I have added/updated functional tests where appropriate
* [ ] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/resources/application.properties [hunk 1]
```diff
@@ -9,3 +9,4 @@ management.endpoints.web.base-path= /users
 management.endpoints.web.path-mapping.health=healthcheck
 management.endpoint.health.enabled=true
 management.endpoint.health.show-details=never
+database.limit=50
```
Comment (bcullerton) [inline src/main/resources/application.properties:12]: We already take a default of 50 from the code. Is it worth making this an optional env var? Gives a bit of configuration in future

File: src/main/java/uk/gov/companieshouse/accounts/user/controller/FindUserBasedOnPartialEmailController.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        ResponseEntity<UsersList> response = new ResponseEntity<>( users, HttpStatus.NO_CONTENT);
+
+        if (! users.isEmpty() ) {
+            users.forEach(user -> userEmailList.append(user.getEmail()).append(" "));
+            LOG.debug( String.format( "%s: Successfully fetched %d users : %s",
+                    xRequestId, 
+                    users.size(), 
+                    userEmailList.toString()));
+            
+            response = new ResponseEntity<>( users, HttpStatus.OK );
```
Comment (bcullerton) [inline src/main/java/uk/gov/companieshouse/accounts/user/controller/FindUserBasedOnPartialEmailController.java:57]: Could just return new ResponseEntity here and in the else. Saves re-assigning NO_CONTENT

File: src/main/java/uk/gov/companieshouse/accounts/user/controller/FindUserBasedOnPartialEmailController.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public ResponseEntity<UsersList> searchUsersDetailsUsingPartialEmail( final String xRequestId, final String partialEmail ) {
+
+        if(Objects.isNull(partialEmail) || partialEmail.isEmpty()){
+            LOG.error(String.format("%s: No partial email was provided.", xRequestId));
+            throw new BadRequestRuntimeException("Please check the request and try again");
+        }
+
+        LOG.debug( 
+            String.format( "%s: Attempting to search for users with an email address containing: %s",   xRequestId, partialEmail)
```
Comment (bcullerton) [inline src/main/java/uk/gov/companieshouse/accounts/user/controller/FindUserBasedOnPartialEmailController.java:37]: Spacing looks a little off here. Double space after first comma?

**Review by bcullerton — COMMENTED** [review]
Looks good, I'd say either update the endpoint-documentation.md or create a new internal-endpoint-documentation.md as part of this ticket.

File: src/main/resources/application.properties (hunk 1 above)
Comment (dtoulcher-ch) [reply src/main/resources/application.properties:12]: I've only added it here is just for improved visibility, I'm happy to remove it

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Review by bcullerton — DISMISSED** [review]

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by hselvakumar — DISMISSED** [review]

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by bcullerton — APPROVED** [review]

**Pull request merged**